### ISSUE TITLE
APIs for service locator/IoC?

### ISSUE THREAD

**Issue by alex-oleshkevich** [Kludex/starlette#416]
What is your opinion about having a service locator API in the framework?
With this, we can use contracts to inject dependencies as well as easily share and replace components.
Also this would be a very nice feature for extension writers as they would be able to independently expose services.

http://code.activestate.com/recipes/413268/
https://fastapi.tiangolo.com/tutorial/dependencies/first-steps/

We also can use type hints for automatic service resolution.
https://laravel.com/docs/5.7/container
https://symfony.com/doc/current/service_container.html

**Comment by lovelydinosaur** [comment]
I think it probably doesn't belong in Starlette's scope. Bocadillo and FastAPI both provide dependency injection over the top of Starlette, which is great.

Somewhat related to this - we'll be moving to recommending the more explicit routing style of...

```python
routes = [
   Route('/', homepage),
   ...
```

Rather than the decorator style...

```python
@app.route('/')
async def homepage(request):
    ...
```

That's relevant here, because we'll actually also make sure to expose good API for overriding the behavior of Routing classes, so that developers that have them handle dispatching differently.
For example...

```python
from some_depenency_injection_package import Route  # Alternates to the standard Starlette routes.

routes = [ 
    Route('/', homepage)  # Dispatch using dependency injection, rather than `func(req) -> resp`.
    ...
]
```

This means that Starlette can stick with it's standard req/resp style interface, but alternative packages or frameworks can easily add other styles. `func(req, resp)` or `func(<dependency injected>) -> resp` or whatever.

In short - not against it, but I think it's a pattern-too-far for what we need Starlette to focus on right now. (So let's provide for making it possible, without having it be in Starlette itself)
